README.md
# snips-skill-respeaker

<<<<<<< HEAD
[![Version](https://img.shields.io/badge/snips--led--animation-v1.6.3-blue.svg)](https://github.com/snipsco/snips-skill-respeaker/blob/master/CHANGELOG.md)
=======
<<<<<<< HEAD
![Version](https://img.shields.io/badge/snips--led--animation-v1.6.3-blue.svg)
=======
[![Version](https://img.shields.io/badge/snips--led--animation-v1.6.2-blue.svg)](https://github.com/snipsco/snips-skill-respeaker/blob/master/CHANGELOG.md)
>>>>>>> 8eb87f718b21c3ed251b2b1ddd8db569388be40b
>>>>>>> 96c09674aebc87b6e0540fec6ac1fd10774bda8a
[![MIT License](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/snipsco/snips-skill-respeaker/blob/master/LICENSE)

This action code animates RGB LEDs and controls button as a human interface for Snips Voice Platform.

User are not only expected to attach a ReSpeaker Hat as the hardware, but also the customised RGB LED (APA102) strip and button are also supported.

## Features

### LED Animations

:bulb: ***```On idle: random led breathe in green```***

:hourglass_flowing_sand: ***```On listen: all leds start to blink in blue```***

:loudspeaker: ***```On speak: all leds start to breathe in purple```***

:mute: ***```To mute: all leds breathe once in yellow```***

:sound: ***```To unmute: all leds breathe once in green```***

:zzz: ***```On disabled: all leds turn off```***

### Button

:star: ***```Short press: start a new conversion without saying hotword```***

:star2: ***```Double press: toggle the animation feedback```***

:dizzy: ***```Long press: toggle the sound feedback```***

## Official Hardware Support

| Platforms / ReSpeakers    |  2-Mic Hat          | 4-Mic Hat           | 6-Mic Hat          | Core V2              |
| :---:                     |  :---:              | :---:               | :---:              | :---:                |
| **Raspberry Pi 0**        |  :white_check_mark: | :white_check_mark:  | :white_check_mark: | :heavy_minus_sign:   |
| **Raspberry Pi 3**        |  :white_check_mark: | :white_check_mark:  | :white_check_mark: | :heavy_minus_sign:   |
| **Raspberry Pi 3+**       |  :white_check_mark: | :white_check_mark:  | :white_check_mark: | :heavy_minus_sign:   |
| **ReSpeaker Core V2**     |  :heavy_minus_sign: | :heavy_minus_sign:  | :heavy_minus_sign: | :white_check_mark:   |

> Known Issue: the button on the ReSpeaker Core V2 is not yet supported due to its special accessing design
>
> If you would like to customise the hardware, please refer to [Build Your Own Hardware](https://github.com/snipsco/snips-skill-respeaker/tree/master#build-your-own-hardware) section.

## Installation

#### By using `sam`:

> Beware that `sam` only supports Raspberry Pi platform. For other platforms, please refer to manual installation guide.

```bash
sam install actions -g https://github.com/snipsco/snips-skill-respeaker.git
```

To be able to access GPIO and SPI hardware, `_snips-skills` user need to be appended with `spi`, `gpio`, `audio` groups. Run the following command on your snips device:

```bash
sudo usermod -a -G spi,gpio,audio _snips-skills
```

#### Manually:

> If there is no `snips-skill-server` installed yet, please run the following command:
>
> `sudo apt-get install snips-skill-server`

Firstly, add your current user to the `snips-skills-admin` group:

```bash
sudo usermod -a -G snips-skills-admin $USER
```

Secondly, add `_snips-skills` user to the `spi`, `gpio`, `audio` groups:

```bash
sudo usermod -a -G spi,gpio,audio _snips-skills
```

Finally, clone the repository and manually run `setup.sh`:

```bash
git clone  https://github.com/snipsco/snips-skill-respeaker.git
cp -r snips-skill-respeaker /var/lib/snips/skills/
cd /var/lib/snips/skills/snips-skill-respeaker
./setup.sh
```

## Configurations

All the configuration options are written in `config.ini` file. There are three sections used to represent three different kinds of parameters. Please refer to your actual usage to modify.

### `[secret]`

This section contains the options which will be asked to input from user during the installation. (Using sam)

#### Device Info

| Config | Description | Value | Default |
| --- | --- | --- | --- |
| `site_id` | Snips device ID | Refering to the actual `snips.toml` | `default` |

> ***To make satellite work correctly, please change here***

#### MQTT

| Config | Description | Value | Default |
| --- | --- | --- | --- |
| `mqtt_host` | MQTT host name | `<ip address>`/`<hostname>` | `localhost` |
| `mqtt_port` | MQTT port number | `<mqtt port>` | `1883` |
| `mqtt_username` | *Optional* - MQTT username | `<mqtt username>` | `null` |
| `mqtt_password` | *Optional* - MQTT password | `<mqtt password>` | `null` |

> ***To make satellite work correctly, please change here***

### `[global]`

This section contains all the changeable options which may make the action more suit your expectation.

#### Hardware Info

| Config | Description | Value | Default |
| --- | --- | --- | --- |
| `model` | Hardware specification file name | Refering to the [hardware_specs](https://github.com/snipsco/snips-skill-respeaker/tree/master/hardware_specs) list | `respeaker_2_mic_hat` |

If you would like to use an APA102 LEDs strip or an external button, please reach [here](https://github.com/snipsco/snips-skill-respeaker#build-your-own-hardware) to create your own hardware configuration file.

#### Brightness

| Config | Description | Value | Default |
| --- | --- | --- | --- |
| `led_bri` | Max brightness of LEDs | `0`:`255` | `64` |

#### Animation

| Config | Description | Value | Default |
| --- | --- | --- | --- |
| `on_idle` | Random LED breathe in green | `true` , `false` | `true` |
| `on_listen` | All LEDs start to blink in blue | `true` , `false` | `true` |
| `on_speak` | All LEDs start to breathe in purple | `true` , `false` | `true` |
| `to_mute` | All LEDs breathe once in orange | `true` , `false` | `true` |
| `to_unmute` | All LEDs breathe once in green | `true` , `false` | `true` |

#### Colour

| Config | Description | Value | Default |
| --- | --- | --- | --- |
| `idle_colour` | LEDs colour while idle | `red` , `green` , `blue` , `yellow` , `purple` , `teal` , `orange` | `green` |
| `listen_colour` | LEDs colour while listening | `red` , `green` , `blue` , `yellow` , `purple` , `teal` , `orange` | `blue` |
| `speak_colour` | LEDs colour while speaking | `red` , `green` , `blue` , `yellow` , `purple` , `teal` , `orange` | `purple` |
| `mute_colour` | LEDs colour while muting | `red` , `green` , `blue` , `yellow` , `purple` , `teal` , `orange` | `yellow` |
| `unmute_colour` | LEDs colour while unmuting | `red` , `green` , `blue` , `yellow` , `purple` , `teal` , `orange` | `green` |

#### Nightmode

| Config | Description | Value | Default |
| --- | --- | --- | --- |
| `nightmode` | If enable night mode | `true` , `false` | `true` |
| `go_sleep` | The time when it sleep | `HH:MM` (24h) | `22:15` |
| `go_wake` | The time when it wake up | `HH:MM` (24h) | `8:35` |

> ***To be able to use this function, please properly set your timezone***
>
> ***Timezone setting: `sudo raspi-config` -> `Localisation Options` -> `Change Timezone`***

#### Default feedback sound

| Config | Description | Value | Default |
| --- | --- | --- | --- |
| `feedback_sound` | If enable the feedback sound by default | `true`:`false` | `true` |

### `[static]`

This section only contains one option, `config_ver`, which is used to track the `config.ini` file version. **You are not supposed to change this value at any time.**

A general update will work basing the old `config.ini` without any problem if there is no change required for adding new config options. But this also means that the config entities might be changed at some point. During the installation/updating, `setup.sh` will always check if the old `config.ini` file meets the latest requirement. If it doesn't, it will be overwrote by a new default config. The old config information will be dropped.

In this case, please do make sure that you have to re-modify the options' value after the installation/updating if needed.

## Build Your Own Hardware

If you would like to have a customised setup, a APA102 IC based RGB LED strip will be needed. (ws2812 will be supported in the future)

### Step 1

Create a file under `hardware_specs` folder, name it with the `custom_` as prefix and end with `.json`:
```
cd hardware_specs
touch custom_<YOUR_HW_CONFIG_NAME>.json
```

Name this file in the correct way will prevent it being tracked by `git` tool, so that the update will not overwrite this specification file.

### Step 2

Fill this file with a json formate specification. This json file should contain the following attributes:

##### Specification Json

| Key | Type | Value |
| --- | --- | --- |
| `led_num` | Integer | Number of LEDs |
| `spi_bus` | Integer | SPI bus number |
| `spi_dev` | Integer | SPI device number |
| `power` | Object | *Optional* - See below |
| `button` | Object | *Optional* - See below |

##### Power Specs

If your LED strip has a power control, specify the GPIO that uses.

| Key | Type | Value |
| --- | --- | --- |
| `gpio_pin` | Integer | Number of the GPIO port |
| `gpio_val` | Integer | Output level, `1` for high and `0` for low |

##### Button Specs

If you have a button connected, specify the GPIO that uses.

> You may need an external pull-up/pull-down resistor connected to your button.

| Key | Type | Value |
| --- | --- | --- |
| `gpio_pin` | Integer | Number of the GPIO port |
| `gpio_val` | Integer | Active level, `1` for high and `0` for low |

If you have a 24 LED strip with a power control on GPIO5, and you would also have a button connected to GPIO17, the following is an example that you need to put to the specification file:

```json
{
    "led_num":24,
    "spi_bus":0,
    "spi_dev":0,
    "power":{
        "gpio_pin":5,
        "gpio_val":1
    },
    "button":{
        "gpio_pin":17,
        "gpio_val":0
    }
}
```

### Step 3

Change the `config.ini`, find the key `model`, give your hardware specification file name with out `.json`.

### Step 4

Relaunch `snips-skill-server`

```
sudo systemctl restart snips-skill-server
```

## Command Line Mode

If you would like to debug your hardware configuration, you can run the animation program with command line parameters. CMD input parameters will have higher priority than the parameters in the `config.ini` file, the missing parameters will be fetched from the `config.ini` file.

```
Usage: ./action-led_animation_x <options> command

Global options:
  -H, --help         this help
  -d, --debug        debug mode
  -v, --verbose      enable verbose mode
  -V, --version      print version of this program

MQTT connection options:
  -h, --mqtt-host    hostname of mqtt broker (default localhost)
  -p, --mqtt-port    port of mqtt broker (default 1883)
  -P, --mqtt-pass    password of mqtt broker
  -U, --mqtt-user    username of mqtt broker

APA102 LED options:
  -N, --led-n        number of leds
  -D, --spi-dev      spi device to use (default /dev/spidev0.0)

Snips device info:
  -S, --site-id      siteId to use (default default)
```

## Enable/Disable LED feedback

Mode change follows hermes procotol.

#### Enable

Topic:

`hermes/feedback/led/toggleOn`

Payload:

| Key | Type | Value |
| --- | --- | --- |
| siteId | String | Site where feedback led will be toggled On |

#### Disable

Topic:

`hermes/feedback/led/toggleOff`

Payload:

| Key | Type | Value |
| --- | --- | --- |
| siteId | String | Site where feedback led will be toggled Off |

## To Do

- [ ] Support WS281x RGB LEDs

## Contributing

Please see the [Contribution Guidelines](https://github.com/snipsco/snips-skill-respeaker/blob/master/CONTRIBUTING.md).

## Copyright

This library is provided by [Snips](https://www.snips.ai) as Open Source software. See [LICENSE](https://github.com/snipsco/snips-skill-respeaker/blob/master/LICENSE) for more information.


CHANGELOG.md
# Changelog

## [1.6.3](#) - 2019-01-28

### Improved

- Check if `_snips-skills` user has the `spi` and `gpio` access permission

## [1.6.2](#) - 2019-01-25

### Added

- Support ReSpeaker 6 mic hat

## [1.6.1](https://github.com/snipsco/snips-skill-respeaker/releases/tag/v1.6.1) - 2019-01-24

### Added

- Support all ReSpeaker hats
- Support ReSepaker Core V2
- Support button interaction
- Support all customized APA102 based light hardware
- Support command line parameter input

### Improved

- LED animation is more stable
- Less CPU consumption
- Extendable structure for implement ws2812 LED driver

### Fixed

- A few small bugs

snipsLed.c
#include "include/snipsLed.h"

short       last_state = 0;
short       state_flag = 1;
short       state = 0;
char        *client_id;

uint32_t    numLEDs;
uint32_t    bitrate = 32000000;
int         fd = -1;
int         sockfd = -1;
uint8_t     *pixels = NULL;
uint8_t     rOffset = 1;
uint8_t     gOffset = 2;
uint8_t     bOffset = 3;

pthread_t global_thread[9];

pthread_t curr_thread;
pthread_t last_thread;

void (*f[9])(const char *)={on_idle, on_listen, on_think, on_speak, to_mute, to_unmute, on_success, on_error, on_off};

int main(int argc, const char *argv[]) 
{
    const char* addr;
    const char* port;
    const char* topic[NUM_TOPIC];
    client_id = generate_client_id();
    short i;
    printf("[Info] Client id is: %s\n", client_id);

    /* get address (argv[1] if present) */
    if (argc > 1) {
        addr = argv[1];
    } else {
        addr = "localhost";
    }

    /* get port number (argv[2] if present) */
    if (argc > 2) {
        port = argv[2];
    } else {
        port = "1883";
    }

    if (argc > 3) {
        if (strcmp(argv[3],"3") == 0){
            numLEDs = 3;
        }else if(strcmp(argv[3],"12") == 0){
            numLEDs = 12;
        }
    }else{
        numLEDs = 3;
    }
    
    

    topic[0] = "hermes/hotword/toggleOff";
    topic[1] = "hermes/asr/startListening";
    topic[2] = "hermes/asr/stopListening";
    topic[3] = "hermes/tts/say";
    topic[4] = "hermes/tts/sayFinished";
    topic[5] = "hermes/hotword/toggleOn";
    topic[6] = "hermes/feedback/sound/toggleOn";
    topic[7] = "hermes/feedback/sound/toggleOff";
    topic[8] = "hermes/nlu/intentNotRecognized";
    topic[9] = "hermes/nlu/intentParsed";


    /* open the non-blocking TCP socket (connecting to the broker) */
    sockfd = open_nb_socket(addr, port);

    if (sockfd == -1) {
        perror("Failed to open socket: ");
        terminate(EXIT_FAILURE, NULL);
    }

    /* setup a client */
    struct mqtt_client client;
    uint8_t sendbuf[2048]; /* sendbuf should be large enough to hold multiple whole mqtt messages */
    uint8_t recvbuf[1024]; /* recvbuf should be large enough any whole mqtt message expected to be received */
    mqtt_init(&client, sockfd, sendbuf, sizeof(sendbuf), recvbuf, sizeof(recvbuf), publish_callback);
    mqtt_connect(&client, client_id, NULL, NULL, 0, NULL, NULL, 0, 400);

    /* check that we don't have any errors */
    if (client.error != MQTT_OK) {
        fprintf(stderr, "error: %s\n", mqtt_error_str(client.error));
        terminate(EXIT_FAILURE, NULL);
    }

    /* start a thread to refresh the client (handle egress and ingree client traffic) */
    pthread_t client_daemon;
    if(pthread_create(&client_daemon, NULL, client_refresher, &client)) {
        fprintf(stderr, "Failed to start client daemon.\n");
        terminate(EXIT_FAILURE, NULL);

    }

    /* subscribe */
    for(i=0;i<NUM_TOPIC;i++){
        mqtt_subscribe(&client, topic[i], 0);
        printf("%s listening for '%s' messages.\n", argv[0], topic[i]);
    }

    // for (i = 0; i < 9; i++)
    // {
    //     pthread_create(&global_thread[i], NULL, f[i], NULL);
    // }

    memset(&global_thread, 0, sizeof(global_thread));

    hw_init();

    printf("LEDs : %d\n", numLEDs);
    printf("Press CTRL-D to exit.\n\n");
    
    /* block */
    while(fgetc(stdin) != EOF);
    
    /* disconnect */
    printf("\n%s disconnecting from %s\n", argv[0], addr);
    sleep(1);

    /* terminate */ 
    terminate(EXIT_SUCCESS, &client_daemon);
}

void terminate(int status, pthread_t *client_daemon)
{
    show(0);
    if (sockfd != -1) close(sockfd);
    if (fd) close(fd), fd = -1;
    if (pixels) free(pixels);
    if (client_daemon != NULL) pthread_cancel(*client_daemon);
    exit(status);
}

void publish_callback(void** unused, struct mqtt_response_publish *published) 
{
    int temp;
    pthread_t fresh_state;
    /* note that published->topic_name is NOT null-terminated (here we'll change it to a c-string) */
    char* topic_name = (char*) malloc(published->topic_name_size + 1);
    memcpy(topic_name, published->topic_name, published->topic_name_size);
    topic_name[published->topic_name_size] = '\0';

    printf("[Received] %s \n", topic_name);

    switch(state){
        case 0:
            if (strcmp(topic_name, "hermes/hotword/toggleOff") == 0)
                last_state = state, state = 1;
            else if (strcmp(topic_name, "hermes/feedback/sound/toggleOff") == 0)
                last_state = state, state = 4;
            else if (strcmp(topic_name, "hermes/feedback/sound/toggleOn") == 0)
                last_state = state, state = 5;
            break;
        case 1:
            if (strcmp(topic_name, "hermes/asr/stopListening") == 0)
                last_state = state, state = 2;
            else if (strcmp(topic_name, "hermes/hotword/toggleOn") == 0)
                last_state = state, state = 0;
            break;
        case 2:
            if (strcmp(topic_name, "hermes/tts/say") == 0)
                last_state = state, state = 3;
            else if (strcmp(topic_name, "hermes/hotword/toggleOn") == 0)
                last_state = state, state = 0;
            break;
        case 3:
            if (strcmp(topic_name, "hermes/tts/sayFinished") == 0)
                last_state = state, state = 0;
            else if (strcmp(topic_name, "hermes/hotword/toggleOn") == 0)
                last_state = state, state = 0;
            break;
        
    }

    printf("[Info] State is changed to %d\n", state);
    if (last_state != state)
        pthread_create(&curr_thread, NULL, f[state], NULL);

    free(topic_name);
}

void* client_refresher(void* client)
{
    while(1) 
    {
        mqtt_sync((struct mqtt_client*) client);
        usleep(100000U);
    }
    return NULL;
}

char* generate_client_id(){
    int i, flag;
    static char id[CLIENT_ID_LEN + 1] = {0};
    srand(time(NULL));
    for (int i = 0; i < CLIENT_ID_LEN; i++){
        flag = rand()%3;
        switch(flag){
            case 0:
                id[i] = rand()%26 + 'a';
                break;
            case 1:
                id[i] = rand()%26 + 'A';
                break;
            case 2:
                id[i] = rand()%10 + '0';
                break;
        }
    }
    return id;
}

void hw_init(){
    int temp;
    pixels = (uint8_t *)malloc(numLEDs * 4);
    begin();
    if((temp = pthread_create(&curr_thread, NULL, on_idle, NULL)) != 0)
        printf("faild to create 1st thread!\n"); 
    else 
        printf("created thread.\n");
}

void begin(){
    if((fd = open("/dev/spidev0.0", O_RDWR)) < 0) {
        printf("[Error] Can't open /dev/spidev0.0 (try 'sudo')");
        terminate(EXIT_FAILURE, NULL);
    }

    uint8_t mode = SPI_MODE_0 | SPI_NO_CS; //SPI_NO_CS   SPI_CS_HIGH

    ioctl(fd, SPI_IOC_WR_MODE, &mode);
    ioctl(fd, SPI_IOC_WR_MAX_SPEED_HZ, &bitrate);
}

void set_color(uint32_t index, uint32_t color){
    uint8_t  r, g, b;

    r = color >> 16;
    g = color >>  8;
    b = color;

    if(index < numLEDs) {
        uint8_t *ptr = &pixels[index * 4];
        ptr[rOffset] = r;
        ptr[gOffset] = g;
        ptr[bOffset] = b;
    }
}

void show(uint8_t brightness){
    if (fd < 0) return;
    if (brightness > 255) brightness = 255;
    if (brightness < 0) brightness = 0;

    uint32_t    i;
    uint8_t     *ptr    =   pixels;
    uint16_t    scale   =   brightness;

    uint8_t x[4];
    // Start Signal:
    x[0] = 0;
    i    = 4;
    while(i--) (void)write(fd, x, 1);

    // Payload:
    x[0] = 0xFF;
    for(i=0; i<numLEDs; i++, ptr += 4) {

        x[1] = (ptr[1] * scale) >> 8;
        x[2] = (ptr[2] * scale) >> 8;
        x[3] = (ptr[3] * scale) >> 8;

        (void)write(fd, x, sizeof(x));
    }

    // Stop Signal:
    x[0] = 0x01;
    //i = (numLEDs + 15) / 16;
    i = 32;
    while(i--) (void)write(fd, x, 1);
}

void clear(){
    uint8_t *ptr;
    uint32_t i;
    for(ptr = pixels, i=0; i<numLEDs; i++, ptr += 4) {
        ptr[1] = 0x05; ptr[2] = 0x05; ptr[3] = 0x05;
    }
    show(5);
}

// States API

// 0
void *on_idle(){
    uint8_t i,j;
    int curr_bri = 0;
    printf("[Thread] ------>  on_idle started\n");
    clear();
    while(state == 0){
        // every 5 seconds
        for(i=0; i<numLEDs; i++){
            if(state != 0) {clear();return((void *)0);}
            set_color(i, 0x00FF00);
        }
        while(curr_bri <= 64){
            // each 0.1s
            if(state != 0) {clear();return((void *)0);}
            show(curr_bri);
            curr_bri += 5;
            for (int j = 0; j < 10; j++){
                // each 0.01s check
                if(state != 0) {clear();return((void *)0);}
                usleep(10000);
            }
        }
        curr_bri = 64;
        while(curr_bri >= 0){
            // each 0.1s
            if(state != 0) {clear();return((void *)0);}
            show(curr_bri);
            curr_bri -= 5;
            for (int j = 0; j < 10; j++){
                // each 0.01s check
                if(state != 0) {clear();return((void *)0);}
                usleep(10000);
            }
        }
        curr_bri = 0;
        for (int j = 0; j < 500; j++){
            // each 0.05s check
            if(state != 0) {clear();return((void *)0);}
            usleep(10000);
        }
    }
    clear();
    return((void *)0);
}

// 1
void *on_listen(){
    uint8_t i,g,group;
    printf("[Thread] ------>  on_listen started\n");
    clear();
    group = numLEDs/3;
    while(state == 1){
        for(i=0; i<3; i++){
            clear();
            for (g=0; g < group; g++)
                set_color(g*3+i, 0xFF0000);
            show(255);
            if(state != 1) {clear();return((void *)1);}
            usleep(100000);
            show(5);
            if(state != 1) {clear();return((void *)1);}
            usleep(100000);
            clear();
        }
    }
    clear();
    return((void *)1);   
}

// 2
void *on_think(){
    uint8_t i,g,group;
    printf("[Thread] ------>  on_think started\n");
    clear();
    group = numLEDs/3;
    while(state == 2){
        for(i=2; i>0; i--){
            clear();
            for (g=0; g < group; g++)
                set_color(g*3+i, 0x00FF00);
            show(255);
            if(state != 2) {clear();return((void *)2);}
            usleep(100000);
            show(5);
            if(state != 2) {clear();return((void *)2);}
            usleep(100000);
            clear();
        }
    }
    clear();
    return((void *)2);
}

// 3
void *on_speak(){
    uint8_t i,g,group;
    printf("[Thread] ------>  on_speak started\n");
    clear();
    group = numLEDs/3;
    while(state == 3){
        for(i=2; i>0; i--){
            clear();
            for (g=0; g < group; g++)
                set_color(g*3+i, 0x00FF00);
            show(255);
            if(state != 3) {clear();return((void *)3);}
            usleep(100000);
            show(5);
            if(state != 3) {clear();return((void *)3);}
            usleep(100000);
            clear();
        }
    }
    clear();
    return((void *)3);
}

// 4
void *to_mute(){
    uint8_t i,j,k;
    int curr_bri = 5;
    printf("[Thread] ------>  to_mute started\n");
    clear();
    clear();
    clear();
    for (int k = 0; k < 2; ++k){
        for(i=0; i<numLEDs; i++){
            if(state != 4) {clear();return((void *)4);}
            set_color(i, 0x0000FF);
        }
        printf("set color done\n");
        while(curr_bri <= 128){
            if(state != 4) {clear();return((void *)4);}
            show(curr_bri);
            curr_bri += 10;
            usleep(80000);
        }
        printf("Turn up done\n");
        curr_bri = 128;
        while(curr_bri >= 0){
            if(state != 4) {clear();return((void *)4);}
            show(curr_bri);
            curr_bri -= 10;
            usleep(80000);
        }
        curr_bri = 0;
        printf("Turn down done\n");  
    }

    last_state = state;
    state = 0;
    clear();
    if (last_state != state)
        pthread_create(&curr_thread, NULL, f[state], NULL);
    return((void *)4);
}

// 5
void *to_unmute(){
    uint8_t i,j;
    int curr_bri = 0;
    printf("[Thread] ------>  to_unmute started\n");
    clear();
    while(state == 5){
        // every 5 seconds
        for(i=0; i<numLEDs; i++){
            if(state != 5) {clear();return((void *)5);}
            set_color(i, 0x00FF00);
        }
        while(curr_bri <= 64){
            if(state != 5) {clear();return((void *)5);}
            show(curr_bri);
            curr_bri += 5;
            usleep(80000);
        }
        curr_bri = 64;
        while(curr_bri >= 0){
            if(state != 5) {clear();return((void *)5);}
            show(curr_bri);
            curr_bri -= 5;
            usleep(80000);
        }
        curr_bri = 0;
        last_state = state;
        state = 0;
    }
    clear();
    if (last_state != state)
        pthread_create(&curr_thread, NULL, f[state], NULL);
    return((void *)5);
}

// 7 
void *on_success(){
    uint8_t i;
    int curr_bri = 0;
    printf("[State] ------>  on_success started\n");

    while(1){
        if (state == 7)
        {
            state = last_state;
            for(i=0; i<numLEDs; i++){
                if(state != 7) break;
                set_color(i, 0x00FF00);
            }
            while(curr_bri <= 255){
                if(state != 7) break;
                show(curr_bri);
                curr_bri += 20;
                usleep(20000);
            }
            curr_bri = 255;
            while(curr_bri >= 0){
                if(state != 7) break;
                show(curr_bri);
                curr_bri -= 20;
                usleep(20000);
            curr_bri = 0;  
            }
        }
    }
}

// 6
void *on_error(){
    uint8_t i;
    int curr_bri = 0;
    printf("[State] ------>  on_error started\n");

    while(1){
        if (state == 6)
        {
            state = last_state;
            for(i=0; i<numLEDs; i++){
                if(state != 6) break;
                set_color(i, 0x0000FF);
            }
            while(curr_bri <= 255){
                if(state != 6) break;
                show(curr_bri);
                curr_bri += 20;
                usleep(20000);
            }
            curr_bri = 255;
            while(curr_bri >= 0){
                if(state != 6) break;
                show(curr_bri);
                curr_bri -= 20;
                usleep(20000);
            curr_bri = 0;  
            }
        }
    }
}

// 7
void *on_off(){
    clear();
    pthread_exit(1);
}


